### PR TITLE
Continuous Deployment ftw

### PR DESCRIPTION
**Pull request by Silvernitro**
### Changes
- Set up Dockerfiles for user-service and frontend (not sure what's going into matching-service yet so we'll revisit this again)
- Set up docker compose to launch user-service, its redis service, and the frontend
- Add a Github Actions script to deploy user-service and frontend anytime code is merged to `develop`
- A few miscellaneous and small changes to user-service and frontend that I'll mark out in comments below

###  How our CD works
Our services are deployed on a Digital Ocean Droplet, which is basically a rented VM. To make things simple to launch for now, I'm just using docker compose so we can launch all our services in just one command. As mentioned, we can look at k8s to deploy them in true microservice spirit next time.

What happens when code is merged to main and the Github Actions script runs:
1. We ask Github Actions to build new Docker images for our services, using the latest code on `develop`
2. Push these images to DigitalOcean's Container Registry (in our team's registry repo that is)
3. Using SSH,
    1. First we copy the latest `docker-compose.yml` to the Droplet from GitHub.
    2. Then, pull the latest Docker images from the DigitalOcean Container Registry to the Droplet, which are the ones we've just built and pushed.
    3. Stop the currently running docker containers and start them again with our latest images.
4. It takes some time to restart the docker compose stack. After that, the new code on `develop` should be live! Visit the [Droplet's IP ](http://157.230.243.78) to verify.

### CODE REVIEW

File: frontend/src/utils/configs.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
 //------- CONNECTION OR SETUP CONFIGS --------//
-const URI_USER_SVC = process.env.URI_USER_SVC || 'http://localhost:8000'
+const URI_USER_SVC =
```
Comment (Silvernitro) [inline frontend/src/utils/configs.js:2]: Environment variables in a react app must start with `REACT_APP` in order for `create-react-app` to read them at build time

File: user-service/controller/user-controller.js [hunk 2]
```diff
@@ -16,7 +16,7 @@ export const getUser = async (req, res) => {
   const { userId } = req
 
   const user = await _findUserById(userId, { password: 0 })
-  if (user.err) {
+  if (!user || user.err) {
```
Comment (Silvernitro) [inline user-service/controller/user-controller.js:19]: Error case I missed out previously. If no user with the given userId exists, the returned result is `null`, so we need to check before doing `user.err`

File: user-service/model/redis-repository.js [hunk 3]
```diff
@@ -1,7 +1,12 @@
 import { createClient } from 'redis'
 import { REDIS_JWT_KEY, JWT_EXPIRY } from '../constants.js'
+import 'dotenv/config'
 
-const client = createClient()
+const REDIS_URI = process.env.REDIS_URI || 'localhost:6379'
```
Comment (Silvernitro) [inline user-service/model/redis-repository.js:5]: Added an optional REDIS_URI env variable. This allows user-service to connect to the redis service running on its local docker network. 

Note that when you're running locally and don't have this env variable set, user-service just connects to your localhost redis server as before.

File: .github/workflows/build_and_deploy.yml [hunk 4]
```diff
@@ -0,0 +1,74 @@
+name: CD
+
+on:
+  push:
+    branches: [main, ci-cd]
```
Comment (Silvernitro) [inline .github/workflows/build_and_deploy.yml:5]: I added a trigger for the `ci-cd` branch for testing purposes. If we need to add more stuff to the CI/CD pipeline next time, we can use this branch to test

**Comment by RyanCheungJF** [comment]
LGTM cd ftw yes

**Pull request merged**